### PR TITLE
Cnv calling using cnmops

### PR DESCRIPTION
**Pull request by TammushB**
This PR contains R scripts related to CNV calling. 
Additionally, added a corresponding conda environment and updated the build_vc_docker script accordingly. 
Next step would be decide how we add tests and wrap these scripts as a module within ugvc.

### CODE REVIEW

File: ugvc/cnv/cnv_calling_using_cnmops.R [hunk 1]
```diff
@@ -0,0 +1,40 @@
+suppressPackageStartupMessages(library(cn.mops))
```
Comment (ilyasoifer) [inline ugvc/cnv/cnv_calling_using_cnmops.R:1]: License

**Review by ilyasoifer — APPROVED** [review]

**Comment by ilyasoifer** [comment]
I think our code has to carry the license (like all pipeline scripts in ugvc do)
________________________________

**Comment by TammushB** [comment]
Yet, these files are not pipeline; they are utilities. 
We can create a pipeline/wrapper that will allow running locally for small files.

File: ugvc/cnv/cnv_calling_using_cnmops.R [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#    CNV calling is done using cn.mops algorithm documented here : https://academic.oup.com/nar/article/40/9/e69/1136601
+# CHANGELOG in reverse chronological order
+
+suppressPackageStartupMessages(library(cn.mops))
+suppressPackageStartupMessages(library(magrittr))
+suppressPackageStartupMessages(library(argparse))
+
+parser <- ArgumentParser()
+parser$add_argument("-cohort_rc", "--cohort_reads_count_file",
+                    help="input cohort reads count file in Rdata format")
```
Comment (gilhornung) [inline ugvc/cnv/cnv_calling_using_cnmops.R:25]: typo: "input cohort read counts file in Rdata format"

File: ugvc/cnv/cnv_calling_using_cnmops.R [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+parser$add_argument("-minWidth", "--min_width_val",
+                    help="min width parameter value",
```
Comment (gilhornung) [inline ugvc/cnv/cnv_calling_using_cnmops.R:27]: Uninformative help

File: ugvc/cnv/cnv_calling_using_cnmops.R [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+cohort_reads_count_file <- args$cohort_reads_count_file
+min_width_val <- args$min_width_val
+cores <- args$parallel
+
+load(cohort_reads_count_file) #variable name : merged_cohort_reads_count
+
+resCNMOPS <- cn.mops(merged_cohort_reads_count,parallel=cores,minWidth = as.integer(min_width_val) )
+resCNMOPS_Int <-calcIntegerCopyNumbers(resCNMOPS)
+
+save(resCNMOPS_Int,file="cohort.cnmops.RData")
```
Comment (gilhornung) [inline ugvc/cnv/cnv_calling_using_cnmops.R:43]: Use saveRDS instead of save()

File: ugvc/cnv/cnv_calling_using_cnmops.R [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+df_cnvs<-as.data.frame(cnvs(resCNMOPS_Int))
+write.csv(df_cnvs,"cohort.cnmops.cnvs.csv", row.names = FALSE,quote=FALSE)
+
+df_cnvr<-as.data.frame(cnvr(resCNMOPS_Int))
+write.csv(df_cnvr,"cohort.cnmops.cnvr.csv", row.names = FALSE,quote=FALSE)
```
Comment (gilhornung) [inline ugvc/cnv/cnv_calling_using_cnmops.R:49]: I see that in the output there is a character CN0, CN1, CN2 etc. Are there any more options? If not, I suggest to save integer values (0, 1, 2 ...) instead. Will be better for downstream applications.

Comment (gilhornung) [reply ugvc/cnv/cnv_calling_using_cnmops.R:49]: Maybe IntegerCopyNumbers does this? Unclear to me

File: ugvc/cnv/get_reads_count_from_bam.R [hunk 6]
```diff
@@ -0,0 +1,26 @@
+suppressPackageStartupMessages(library(cn.mops))
+suppressPackageStartupMessages(library(magrittr))
+suppressPackageStartupMessages(library(argparse))
+
+parser <- ArgumentParser()
+
+parser$add_argument("-i", "--input_bam_file",
+                    help="input bam file path")
+parser$add_argument("-refseq", "--refSeqNames_string",
+                    help="chromosomes names in comma seperated format. e.g, chr1,chr2,chrX" ,
```
Comment (gilhornung) [inline ugvc/cnv/get_reads_count_from_bam.R:10]: Typo: chromosome names

File: ugvc/cnv/get_reads_count_from_bam.R [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+                    default="chr1,chr2,chr3,chr4,chr5,chr6,chr7,chr8,chr9,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19,chr20,chr21,chr22,chrX,chrY")
+parser$add_argument("-wl", "--window_length",
+                   help="window length (#bp) for which reads count is calculated for",
+                   type="integer", default=1000)
+parser$add_argument("-p", "--parallel",
+                    help="number of parallel process",
```
Comment (gilhornung) [inline ugvc/cnv/get_reads_count_from_bam.R:16]: Typo: processes

File: ugvc/cnv/merge_reads_count_sample_to_cohort.R [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                    help="input cohort reads count file in Rdata format")
+parser$add_argument("-sample_rc", "--sample_reads_count_file",
+                    help="input sample reads count file in Rdata format")
+args <- parser$parse_args()
+
+cohort_reads_count_file <- args$cohort_reads_count_file
+sample_reads_count_file <- args$sample_reads_count_file
+
+# load reads count Rdata file
+RC_obj<-load(cohort_reads_count_file)
```
Comment (gilhornung) [inline ugvc/cnv/merge_reads_count_sample_to_cohort.R:16]: Prefer rds over Rdata

File: ugvc/cnv/get_reads_count_from_bam.R [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                    type="integer", default=30)
+parser$add_argument("-o", "--base_file_name",
+                    help="out base file name")
+
+args <- parser$parse_args()
+
+refSeqNames <- unlist(strsplit(args$refSeqNames_string, ","))
+bamDataRanges_RC <- getReadCountsFromBAM(args$input_bam_file, refSeqNames=refSeqNames, WL=args$window_length ,parallel=args$parallel)
+save(bamDataRanges_RC, file=paste(args$base_file_name,".ReadCounts.Rdata",sep = ""))
+write.csv(as.data.frame(bamDataRanges_RC),paste(args$base_file_name,".ReadCounts.csv",sep=""),row.names = FALSE)
```
Comment (gilhornung) [inline ugvc/cnv/get_reads_count_from_bam.R:26]: Don't think you need to a csv file. We already have the bigwigs from the pipeline and can work with them

File: ugvc/cnv/get_reads_count_from_bam.R [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+                    help="number of parallel process",
+                    type="integer", default=30)
+parser$add_argument("-o", "--base_file_name",
+                    help="out base file name")
+
+args <- parser$parse_args()
+
+refSeqNames <- unlist(strsplit(args$refSeqNames_string, ","))
+bamDataRanges_RC <- getReadCountsFromBAM(args$input_bam_file, refSeqNames=refSeqNames, WL=args$window_length ,parallel=args$parallel)
+save(bamDataRanges_RC, file=paste(args$base_file_name,".ReadCounts.Rdata",sep = ""))
```
Comment (gilhornung) [inline ugvc/cnv/get_reads_count_from_bam.R:25]: Use saveRDS() instead of save()

File: ugvc/cnv/merge_reads_count_sample_to_cohort.R [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+merged_cohort_reads_count <- GRanges(
+  seqnames = seqnames(gr1),
+  ranges = ranges(gr1),
+  strand = strand(gr1),
+  mcols(gr1),
+  mcols(gr2)
+)
+
+# save merged cohort
+save(merged_cohort_reads_count,file="merged_cohort_reads_count.RData")
```
Comment (gilhornung) [inline ugvc/cnv/merge_reads_count_sample_to_cohort.R:34]: saveRDS() instead of save()

File: ugvc/cnv/merge_reads_count_sample_to_cohort.R [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+write.csv(as.data.frame(merged_cohort_reads_count),"merged_cohort_reads_count.csv", row.names = FALSE)
```
Comment (gilhornung) [inline ugvc/cnv/merge_reads_count_sample_to_cohort.R:35]: merged_cohort_reads_count.csv is a huge file and will probably not be used

**Review by gilhornung — CHANGES_REQUESTED** [review]
Major comments:
1) Remove some large csv files which will probably be never used.
2) Large csv files that will be used - save in HDF5 format instead of csv. This way or the other we will use pandas to load them. It is faster (and cheaper) to store and load in this format (small csv files can be kept in csv).

Minor comments:
1) Use RDS files instead of Rdata files. They are easier to work with, because you do not load an environment, just one object.
2) Typos

File: ugvc/cnv/get_reads_count_from_bam.R (hunk 9 above)
Comment (TammushB) [reply ugvc/cnv/get_reads_count_from_bam.R]: changed to hdf5 files

File: ugvc/cnv/merge_reads_count_sample_to_cohort.R (hunk 12 above)
Comment (TammushB) [reply ugvc/cnv/merge_reads_count_sample_to_cohort.R]: changed to hdf5 file.

File: ugvc/cnv/cnv_calling_using_cnmops.R (hunk 5 above)
Comment (TammushB) [reply ugvc/cnv/cnv_calling_using_cnmops.R]: saved all outputs into a single hdf5.

**Comment by TammushB** [comment]
@gilhornung thanks for all informative comments. updated all outputs to RDS and hdf5 data types. 
@ilyasoifer please review filtering script with corresponding test. 
Next step will be tests for R scripts.

**Review by gilhornung — APPROVED** [review]
No comments

File: ugvc/cnv/filter_sample_cnvs.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+    os.system(cmd)
+    out_len_annotate_file = prefix + out_filename.rstrip(".bed") + ".len.annotate.bed"
+    cmd = "cat " + out_len_file + ' | awk \'{print $1"\t"$2"\t"$3"\tLEN"}\' > ' + out_len_annotate_file
+
+    os.system(cmd)
+    return out_len_annotate_file
+
+
+def annotate_bed(bed_file, lcr_cutoff, lcr_file, diff_cutoff, diff_bed_file, prefix, length_cutoff=10000):
```
Comment (ilyasoifer) [inline ugvc/cnv/filter_sample_cnvs.py:103]: If I understand correctly, filter_by_lcr and filter_by_blocklist is pretty much the same function that is pretty generic for BED files. Can it be written as such and live in vcfbed directory?

**Review by ilyasoifer — COMMENTED** [review]
Hi, 
good job, some comments, happy to discuss!

File: ugvc/cnv/filter_sample_cnvs.py (hunk 13 above)
Comment (TammushB) [reply ugvc/cnv/filter_sample_cnvs.py:13]: Done

File: ugvc/cnv/filter_sample_cnvs.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        args.input_bed_file,
+        args.intersection_cutoff,
+        args.coverage_lcr_file,
+        args.intersection_cutoff,
+        args.blocklist,
+        prefix,
+        args.min_cnv_length,
+    )
+
+    print("output files:")
```
[the remark this answers is not in the document]
Comment (TammushB) [reply ugvc/cnv/filter_sample_cnvs.py]: Done.

**Comment by TammushB** [comment]
Resolved all comments. merging this.

**Pull request merged**